### PR TITLE
feat:[SSCA-610]: Added Debezium Service for InstanceNG Collection Replication

### PR DESCRIPTION
**Pull request by arpit-harness**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/53641)

### CODE REVIEW

File: ssca-manager/service/src/main/java/io/harness/ssca/services/ArtifactServiceImpl.java [hunk 1]
```diff
@@ -116,6 +129,22 @@ public Optional<ArtifactEntity> getArtifact(
...
+      String accountId, String orgIdentifier, String projectIdentifier, String artifactCorrelationId) {
+    Criteria criteria = Criteria.where(ArtifactEntityKeys.accountId)
+                            .is(accountId)
+                            .and(ArtifactEntityKeys.orgId)
+                            .is(orgIdentifier)
+                            .and(ArtifactEntityKeys.projectId)
+                            .is(projectIdentifier)
+                            .and(ArtifactEntityKeys.artifactId)
+                            .is(artifactCorrelationId)
+                            .and(ArtifactEntityKeys.artifactCorrelationId)
```
Comment (inderschera) [inline ssca-manager/service/src/main/java/io/harness/ssca/services/ArtifactServiceImpl.java:143]: Check the query once

File: ssca-manager/service/src/main/java/io/harness/ssca/services/CdInstanceSummaryServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+@Slf4j
+public class CdInstanceSummaryServiceImpl implements CdInstanceSummaryService {
+  @Inject CdInstanceSummaryRepo cdInstanceSummaryRepo;
+
+  @Inject ArtifactService artifactService;
+
+  @Override
+  public boolean upsertInstance(Instance instance) {
+    if (Objects.isNull(instance.getPrimaryArtifact().getArtifactIdentity())
```
Comment (inderschera) [inline ssca-manager/service/src/main/java/io/harness/ssca/services/CdInstanceSummaryServiceImpl.java:35]: instance.getPrimaryArtifact() add null check on this as well

File: ssca-manager/service/src/main/java/io/harness/ssca/services/CdInstanceSummaryServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+        instance.getPrimaryArtifact().getArtifactIdentity().getImage(), instance.getEnvIdentifier());
+
+    if (Objects.nonNull(cdInstanceSummary)) {
+      cdInstanceSummary.getInstanceIds().add(instance.getId());
+      cdInstanceSummaryRepo.save(cdInstanceSummary);
+    } else {
+      ArtifactEntity artifact =
+          artifactService.getArtifactByCorrelationId(instance.getAccountIdentifier(), instance.getOrgIdentifier(),
+              instance.getProjectIdentifier(), instance.getPrimaryArtifact().getArtifactIdentity().getImage());
+      if (Objects.isNull(artifact)) {
```
Comment (inderschera) [inline ssca-manager/service/src/main/java/io/harness/ssca/services/CdInstanceSummaryServiceImpl.java:51]: Add a skipping log here

File: ssca-manager/service/src/main/java/io/harness/ssca/services/ArtifactServiceImpl.java [hunk 4]
```diff
@@ -238,21 +295,48 @@ private List<ArtifactListingResponse> getArtifactListingResponses(List<ArtifactE
...
+                                              .is(accountId)
+                                              .and(CdInstanceSummaryKeys.orgIdentifier)
+                                              .is(orgIdentifier)
+                                              .and(CdInstanceSummaryKeys.projectIdentifier)
+                                              .is(projectIdentifier);
+    List<CdInstanceSummary> cdInstanceSummaries = cdInstanceSummaryRepo.findAll(artifactDeploymentCriteria);
+
+    Map<String, List<CdInstanceSummary>> artifactDeploymentMap = cdInstanceSummaries.stream().collect(
+        Collectors.groupingBy(cdInstanceSummary -> cdInstanceSummary.getArtifactCorrelationId()));
+
```
Comment (inderschera) [inline ssca-manager/service/src/main/java/io/harness/ssca/services/ArtifactServiceImpl.java:312]: Let's instead create aggregation query on mongo itself.
aggregation contains match on accountId, org, project, and correlationIds.
Then a group by with correlationId, where for each correlationId we just get the count of prod and non prod envs. Check $cond to add conditions before taking the sum.

File: ssca-manager/service/src/main/java/io/harness/ssca/services/ArtifactServiceImpl.java (hunk 1 above)
Comment (arpit-harness) [reply ssca-manager/service/src/main/java/io/harness/ssca/services/ArtifactServiceImpl.java:143]: Done.

File: ssca-manager/service/src/main/java/io/harness/ssca/services/ArtifactServiceImpl.java (hunk 4 above)
Comment (arpit-harness) [reply ssca-manager/service/src/main/java/io/harness/ssca/services/ArtifactServiceImpl.java:312]: this is a small set of records which will not have much effect on the performance. can't we prioritise in-memory grouping over an aggregate query.

File: ssca-manager/service/src/main/java/io/harness/ssca/services/CdInstanceSummaryServiceImpl.java (hunk 2 above)
Comment (arpit-harness) [reply ssca-manager/service/src/main/java/io/harness/ssca/services/CdInstanceSummaryServiceImpl.java:35]: Done.

File: ssca-manager/service/src/main/java/io/harness/ssca/services/CdInstanceSummaryServiceImpl.java (hunk 3 above)
Comment (arpit-harness) [reply ssca-manager/service/src/main/java/io/harness/ssca/services/CdInstanceSummaryServiceImpl.java:51]: Done.

**Review by inderschera — APPROVED** [review]

**Comment by arpit-harness** [comment]
trigger uts

**Comment by arpit-harness** [comment]
trigger uts

**Pull request closed without merging**